File: JPP/Lista3/zad3/finite.py
class FiniteField:
    characteristic = 1234567891

    def __init__(self, val=0):
        self.value = self.to_range(val)

    def to_range(self, x):
        x %= self.characteristic
        while x < 0:
            x += self.characteristic
        return x
    
    def invert(self, x):
        for i in range(self.characteristic):
            if (x * i) % self.characteristic == 1:
                return i
        return -1

    @classmethod
    def get_characteristic(cls):
        return cls.characteristic
    
    def __eq__(self, other):
        return self.value == other.value

    def __ne__(self, other):
        return not self == other

    def __le__(self, other):
        return self.value <= other.value
    
    def __ge__(self, other):
        return self.value >= other.value
    
    def __lt__(self, other):
        return self.value < other.value

    def __gt__(self, other):
        return self.value > other.value

    def __add__(self, other):
        return FiniteField(self.value + other.value)

    def __sub__(self, other):
        return FiniteField(self.value - other.value)

    def __mul__(self, other):
        return FiniteField(self.value * other.value)

    def __truediv__(self, other):
        if other.value == 0:
            raise ValueError("Error: Dividing by zero")
        inv = self.invert(other.value)
        if inv == -1:
            raise ValueError("Error: Irreversible element")
        return FiniteField((self.value * inv) % self.characteristic)

    def __iadd__(self, other):
        self.value = self.to_range(self.value + other.value)
        return self

    def __isub__(self, other):
        self.value = self.to_range(self.value - other.value)
        return self

    def __imul__(self, other):
        self.value = self.to_range(self.value * other.value)
        return self

    def __itruediv__(self, other):
        self.value = (self / other).value
        return self

    def __str__(self):
        return str(self.value)

    def __repr__(self):
        return str(self.value)

# if __name__ == "__main__":
#     a = FiniteField(5)
#     b = FiniteField(1234576)
#     c = FiniteField()
#     print("Given 'a':", a, "Given 'b':", b)
#     print("Characteristic:", FiniteField.get_characteristic())

#     print("a == b ?", a == b)
#     print("a != b ?", a != b)
#     print("a <= b ?", a <= b)
#     print("a >= b ?", a >= b)
#     print("a < b ?", a < b)
#     print("a > b ?", a > b)

#     c = a + b
#     print("a + b =", c)

#     c = a - b
#     print("a - b =", c)

#     c = a * b
#     print("a * b =", c)
#     c = a / b
#     print("a / b =", c)

#     c += a
#     print("c += a :", c)
#     c -= a
#     print("c -= a :", c)
#     c *= a
#     print("c *= a :", c)
#     c /= a
#     print("c /= a :", c)

#     a = FiniteField(int(input("Enter a value for a: ")))
#     print("New value for a:", a)

File: JPP/Lista2/Zad1/finite.cpp
#include <stdio.h>
#include <stdexcept>
#include <iostream>

class FiniteField{
private:
    static const int characteristic = 1234577;

    int value;

    int to_range(int x) const {
        x %= characteristic;
        
        while (x < 0) {
            x += characteristic;
        }

        return x;
    }

    int invert(int x) const {
        for(int i = 0; i < characteristic; i++) {
            if((long long)x*i % characteristic == 1) {
                return i;
            }
        }
        return -1;
    }

public:
    FiniteField() : value(0) {}

    explicit FiniteField(int val) : value(to_range(val)) {}

    static const int get_characteristic() {
        return characteristic;
    }

    bool operator==(const FiniteField& other) const {
        return value == other.value;
    }

    bool operator!=(const FiniteField& other) const {
        return !(*this == other);
    }

    bool operator<=(const FiniteField& other) const {
        return value <= other.value;
    }

    bool operator>=(const FiniteField& other) const {
        return value >= other.value;
    }

    bool operator<(const FiniteField& other) const {
        return value < other.value;
    }

    bool operator>(const FiniteField& other) const {
        return value > other.value;
    }

    FiniteField operator+(const FiniteField& other) const {
        return FiniteField(value + other.value);
    }

    FiniteField operator-(const FiniteField& other) const {
        return FiniteField(value - other.value);
    }

    FiniteField operator*(const FiniteField& other) const {
        return FiniteField(value * other.value);
    }

    FiniteField operator/(const FiniteField& other) const {
        if(other.value == 0) {
            throw std::runtime_error("Error: Dividing by zero");
        }
        
        int inv = invert(other.value);
        if(inv == -1) {
            throw std::runtime_error("Error: Irreversible element");
        }
        return FiniteField((long long)value*inv % characteristic);
    }

    FiniteField& operator=(const FiniteField& other) {
        value = other.value;
        return *this;
    }

    FiniteField& operator+=(const FiniteField& other) {
        value = to_range(value + other.value);
        return *this;
    }

    FiniteField& operator-=(const FiniteField& other) {
        value = to_range(value - other.value);
        return *this;
    }

    FiniteField& operator*=(const FiniteField& other) {
        value = to_range(value * other.value);
        return *this;
    }

    FiniteField& operator/=(const FiniteField& other) {
        value = to_range((*this / other).value);
        return *this;
    }

    friend std::ostream& operator<<(std::ostream& os, const FiniteField& field) {
        return os << field.value;
    }

    friend std::istream& operator>>(std::istream& is, FiniteField& field) {
        return is >> field.value;
        //return is;
    }
};

int main(){
    FiniteField a(5);
    FiniteField b(1234576);
    FiniteField c(0);
    std::cout << "Given 'a': " << a << " Given 'b': " << b << std::endl;
    std::cout << "Characteristic: " << FiniteField::get_characteristic() << std::endl;

    std::cout << "a == b ? " << std::endl;
    if(a == b){
        std::cout << "\ta equals b" << std::endl;
    }
    else{
        std::cout << "\ta is not equal b" << std::endl;
    }

    std::cout << "a != b ? " << std::endl;
    if(a != b){
        std::cout << "\ta is not equal b" << std::endl;
    }
    else{
        std::cout << "\ta equals b" << std::endl;
    }

    std::cout << "a <= b ? " <<  std::endl;
    if(a <= b){
        std::cout << "\ta is less or equal than b" << std::endl;
    }
    else{
        std::cout << "\ta is bigger than b" << std::endl;
    }

    std::cout << "a >= b ? " << std::endl;
    if(a >= b){
        std::cout << "\ta is bigger or equal b" << std::endl;
    }
    else{
        std::cout << "\ta is less than b" << std::endl;
    }

    std::cout << "a < b ? " << std::endl;
    if(a < b){
        std::cout << "\ta is less than b" << std::endl;
    }
    else{
        std::cout << "\ta is bigger or equal b" << std::endl;
    }

    std::cout << "a > b ? " << std::endl;
    if(a > b){
        std::cout << "\ta is bigger than b" << std::endl;
    }
    else{
        std::cout << "\ta is less of equal b" << std::endl;
    }

    c = a + b;
    std::cout << "a + b = " << c << std::endl;

    c = a - b;
    std::cout << "a - b = " << c << std::endl;

    c = a * b;
    std::cout << "a * b = " << c << std::endl;

    c = a / b;
    std::cout << "a / b = " << c << std::endl;

    c += a;
    std::cout << "c += a : " << c << std::endl;

    c -= a;
    std::cout << "c -= a : " << c << std::endl;

    c *= a;
    std::cout << "c *= a : " << c << std::endl;

    c /= a;
    std::cout << "c /= a : " << c << std::endl;

    std::cout << "Enter a value for a: ";    
    std::cin >> a;
    std::cout << "New value for a: " << a << std::endl;

    return 0;
}